src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -239,17 +236,76 @@ def _get_source_file(obj_dir: str, obj_file: str) -> str:
     return src_file
 
 
+def _upload_genomes_to_workspace(
+    gfu: GenomeFileUtil,
+    workspace_id: int,
+    load_id: str,
+    ws_obj_tuples: list[WSObjTuple],
+    data_dir: str,
+) -> list[UploadResult]:
+    """
+    Upload genbank files to the target workspace as Genome in batch. The bulk method fails
+    and an error will be thrown if any of the genome files in batch fails to upload.
+    The order of elements in the returned list corresponds to the order of `ws_obj_tuples`.
+    """
+
+    inputs = [
+        {
+              # GFU has the capability to discern and retrieve the accurate GenBank file from the directory
+            "file": {'path': obj_tuple.container_internal_file_dir},
+            "genome_name": obj_tuple.obj_name,
+            "metadata": {"load_id": load_id},
+        }
+        for obj_tuple in ws_obj_tuples
+    ]
+
+    results = gfu.genbanks_to_genomes(
+        {"workspace_id": workspace_id, "inputs": inputs}
+    )["results"]
+
+    upload_results = []
+    for result_dict, obj_tuple in zip(results, ws_obj_tuples):
+
+        genome_obj_info = result_dict["genome_info"]
+        assembly_obj_info = result_dict["assembly_info"]
+
+        genome_tuple = obj_tuple
+
+        # copy assembly file to the source directory
+        assembly_path = Path(result_dict["assembly_path"])
+        # remove prefix of the assembly_path
+        assembly_path = assembly_path.relative_to(_JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER)
+        assembly_path_local = Path(data_dir) / assembly_path
+        if not os.path.exists(assembly_path_local):
+            raise ValueError(f"Assembly file {assembly_path_local} does not exist")
+
+        source_dir = Path(genome_tuple.host_file_dir)
+        loader_helper.create_hardlink_between_files(source_dir / assembly_path.name, assembly_path_local)
```

Comment:
this one links FASTA file from job_dir into source/NCBI/None/ dir. The one in `_process_genome_objects` links FASTA file to source/WS/env/UPA dir.

Reply:
It's hard to keep track of all the directories...

Is there some way of figuring out, from the contents of the WS/env/UPA dir, where the data came from in the source/NCBI... dir?

Reply:
no. I don't think so.

Reply:
I guess it doesn't matter since the genbank file will be uploaded to the WS at that point, right?

Reply:
yea. At this point, GFU has already returned. And we want to process the FASTA file created by GFU.

Reply:
Actually, we should be able to figure it out from the fasta file name / object name, right?

Reply:
I am not following you. What to compute from fasta file name?

Reply:
The source of the contents in the WS/env/UPA dir

Reply:
hmmm. I still don't know what you want to change.

Reply:
> The source/WS/env/UPA dir are then softlinked to `collectionssource/GTDB/<collection_id>`

I think we said this isn't true right, since the NONE environments have the uploaded.yml file and the UPA dir has the meta.yaml files?

Reply:
The other thing we said was that if the workspace object was uploaded by the workspace uploader and not some other process, then you could figure out the `collectionsource/NONE/GTDB/<source_ver>` dir from the assembly object name of the uploaded data

Reply:
👍

Reply:
:+1: